python-frontend/src/main/java/org/sonar/python/semantic/v2/types/TrivialTypeInferenceVisitor.java

```diff
@@ -359,6 +361,10 @@ private static PythonType resolveTypeAnnotationType(TypeAnnotation typeAnnotatio
       var attributes = new ArrayList<PythonType>();
       attributes.add(new ObjectType(elementsType, TypeSource.TYPE_HINT));
       return new ObjectType(subscriptionExpression.object().typeV2(), attributes, new ArrayList<>(), TypeSource.TYPE_HINT);
+    } else if (expression instanceof BinaryExpression binaryExpression) {
+      var left = resolveTypeAnnotationExpressionType(binaryExpression.leftOperand());
+      var right = resolveTypeAnnotationExpressionType(binaryExpression.rightOperand());
+      return UnionType.or(left, right);
```

Comment:
Should it be `UnionType` or `ObjectType[UnionType[...]]`?

I don't remember exactly what we decided, however I have the feeling it should be the latter (that way any `value` related stuff can go in the `ObjectType` itself).

Reply:
I think it is not that clear cause potentially you can have the:
```python
def foo4(a: list[str] | set[int]):
```